### ISSUE TITLE
Matlab alpha implementation

### ISSUE THREAD

**Issue by federicoparra** [CalculatedContent/WeightWatcher#58]
Hi, I would like to implement alpha within the Deep Learning Matlab framework for early stopping without requiring a validation sample. 

Could you please point me toward or write down the pseudo code to calculating alpha in that context so that I can implement it in Matlab ? 

Thank you!
Federico

**Comment by charlesmartin14** [comment]
WeightWatcher is a simple python program and you should be able to call it directly from MatLab
I have not tried this myself so perhaps there some hiccups but if we can identify them I imagine we could adapt WeightWatcher to allow it to be called easily from MatLab

**Comment by federicoparra** [comment]
oh I'll try this! 
Suppose I have a 2d conv layer's weights as a height*width*channels*filters 4D matrix, what would be the python instruction to get the alpha value (the one that, in the Github instructions, is said it should be used as an early stopping indicator when it falls under 2).

**Comment by charlesmartin14** [comment]
Just run the analyze() function and look at the details dataframe
Each layer will be assigned an alpha

You can send me the results and I can help you to interpret the results

**Comment by federicoparra** [comment]
I'm stuck - I can perfectly call weightwatcher from matlab using a pytorch model such as squeezenet using the following simple matlab statements and get the summary alpha as needed per required early stopping implementation: 

`model = py.torchvision.models.squeezenet1_0(pretrained=true);
ww=py.weightwatcher.WeightWatcher(model);
details = ww.analyze();
summary = struct(ww.get_summary(details));
summary.alpha`

But, if instead of a pytorch model I attempt to pass on a MATLAB model (either dagNetwork, or layerGraph, both of MATLAB's deep learning model types) it produces a MATLAB error "Error using py.weightwatcher.WeightWatcher
Conversion of MATLAB 'DAGNetwork' to Python is not supported." because of course MATLAB doesn't know how to translate to pytorch. 

I looked around and a long shot would be to use MATLAB's exportONNXNetwork function to export to hard drive then import to pytorch or keras. But 1) neither keras nor pytorch have native ways of doing the import so third-party libraries have to be used (that's ok but) 2) saving to disk a model then importing to pytorch/keras is too slow to be used during an iterative process within a learning protocol to do early stopping - it would make the learning so slow every time it needs to write to hard drive / read from hard drive an entire model. 

At this point, given I only need the alpha calculation per layer to then calculate the alpha mean, could you please offer the pseudo code (or point me to it) to calculate alpha out of a weight matrix from a 2d conv layer and perhaps also from a fully connected layer? 

I could probably write a fast MATLAB function to compute that at each iteration to validate an early stopping schedule without introducing much delay at each iteration.

thank you so much!

**Comment by federicoparra** [comment]
Or maybe, if there is a way to call your program so that I can give it a weight matrix from a specific layer and it produces the alpha value, then it would be easy for me to build a MATLAB function that reads layer by layer the weight matrix, formats it into python matrix, sends it over to WW, then gets alpha back for that specific layer, then averages all alphas. That should be very fast actually in terms of processing time.

**Comment by federicoparra** [comment]
also, sorry to bother you, maybe there is a way to call WW for getting the summary alpha faster? I mean, the way I'm calling it I'm sure it is processing a whole bunch of other metrics that are not needed - because I notice that the file writing/reading/reformating is quite fast and that the bottleneck in my code is the actual call to WW for the analysis/summary.

**Comment by federicoparra** [comment]
Also, I notice that in the context of transfer learning, if I import a model (say, mobilenet), "freeze" all its learning layers (learning rate of said layers set to 0) except for the last few for fine tuning, and then run WW every few epochs to verify average (summary) alpha level, I always get the same value, even though the network is clearly learning (the last few layers are learning) and even though training / validation loss values are both decreasing. Is this a bug, or just the expected result given that 99% of the networks' weights are not changing during learning? If the latter is the case, would it make sense to monitor only the alpha of the layers that are actually learning, in my case just the last one?

**Comment by federicoparra** [comment]
@charlesmartin14 

Hi, I'm trying to implement this in Matlab. 

For now, my pseudo code looks like this:

For Conv2d layers:

1. outer loop among filters/kernels
2. inner loop among channels
3. get weight matrix corresponding to this filter number / channel number 
4. do weight matrix * transposed weight matrix
5. get eigenvalues of the resultant matrix
6. append all eigenvalues of all filters(kernels)/channels in one single long vector
7. use python Powerlaw fit function to get alpha from that vector (pass maximum eigenvalue as max value)

I repeat those steps for each Conv2d layer, and then I can average the alphas to get average alpha - for some cases I tested, my obtained average alpha is the same as the one obtained with WeightWatcher tool in python. 

Couple of questions: 
1) how should I deal with fully connected layers, whereas the weights are not a matrix but a vector per neuron? Concretely, of a vector of weights, how should I calculate the correlation matrix to then go ahead to calculate eigenvalues etc ?
2) am I doing everything right? is the "average alpha" just a regular average of the alphas of all layers, as I'm doing?

Thank you so much

everything is almost native in Matlab now, except for the power law function I'm still using the python one because I tried a few implementations of MLE of power law in Matlab and the exponents found where nothing like the ones found by the python library for some reason - but the python library call is fast anyways.

Thank you so much!
Federico

**Comment by charlesmartin14** [comment]
**To compute selected layers,** just specify the layers ids as, say

`ww.analyze( ...,  layers=[1,4,6] )`

You can find the layer ids by first running 

`ww.describe( ... )`

**using WeightWatcher with MatLab**
I do not have the bandwidth to support porting the code to MatLab at this time.
There are many details to be considered, and it is a primary goal of ours to maintain the python codebase in a way that all of our research results are 100% reproducible.

I suggest simply adding support for the MatLab  DAGNetwork format.
This only requires the ability to 
- iterate over the DAGNetwork layers
- extract the weight matrix or Conv2D tensor from each DAGNetwork layer


**Speeding up the WW alpha**

I don't think it will help to reimplement the code in MatLab because the python Libraries uses the same underlying BLAS libraries, and the PowerLaw fits use a brute force search (which can be slow). 

I may be able to speed up the code in the next release, although this would require a bit of work.
If you want to help here, that would be great

**Comment by federicoparra** [comment]
Thank you for your detailed answer.

You missed my currently most important question, which is how do you deal with 1x1 Conv2d layers (aka network in network layers) and how do you deal with fully connected / dense layers? I've read your code but since I'm not fluent in python I don't manage to see through. 

Speed: you are right, the bottleneck is the call to PowerLaw. If I manage to find a way to fit the powerlaw with Matlab tools (I'm reading the module author's paper) I'll time Matlab vs. his python implementation and let you know. 

About WW being able to calculate alpha / other metrics for a given weight matrix, I just think that's another way of making the library flexible, for example in my case I could iterate through DAGnetwork layers within Matlab (which is much easier than to code a python code to do that given I don't code in Python) and just call WW iteratively to get alpha/metrics for each weight matrix - in other words all the mechanics of providing the right weight matrixes would be on the user's side when using the library in this low-level fashion. Good side of it is then anyone can call it from any framework without having to explicitly support said framework. 

Thanks again,
Federico

**Comment by charlesmartin14** [comment]
_You missed my currently most important question, which is how do you deal with 1x1 Conv2d layers_

We don't support layers that only have a vector of weights; there is no alpha here

_Speed: you are right, the bottleneck is the call to PowerLaw._ 
That is exactly what I expected.  Fitting alpha properly is tricky.
Speeding this up, and getting consistent results, will take some time.

_WW being able to calculate alpha / other metrics for a given weight matrix,_ 
This will not solve the performance issue.  
I'll take it into consideration for a future release.

**Comment by charlesmartin14** [comment]
closing this issue
